File: Python/incrementalSearch.py
from function import f
import numpy as np

def incrementalSearch (start, step, stop):
    decimals = 4
    return_list = []
    evaluated = f(start)
    return_list.append({
            'iter':0,
            'x': start,
            'f(x)' : evaluated,
            'root': 'NA',
            'f(a)*f(b)': 'NA',
            'range':'NA'
            })
    points = np.arange(start+step,stop+step,step)
    count = 1
    for point in points:
        mult = float(evaluated * f(point))
        root = True if mult < 0 else False 
        
        row = {
            'iter':count,
            'x': point,
            'f(x)' : round(f(point), decimals),
            'root': root,
            'f(a)*f(b)':round(mult,decimals),
            'range':[round(point-step,decimals),round(point,decimals)]
            }
        count = count + 1
        return_list.append(row)
        evaluated = f(point)

    return {"iters": return_list}

File: C++/incrementalSearch.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stdlib.h>
#include "function.h"

using namespace std;

vector<vector<double>> incrementalSearch(double start, double step, double stop) {
    vector<vector<double>> returnList;
    vector<double> row;
    
    double evaluated_f = f(start);
    row.push_back(start);
    row.push_back(evaluated_f);
    row.push_back(0);
    row.push_back(0);
    returnList.push_back(row);
    row.clear();

    for(double i = start+step; i <= stop; i += step){
        double mult = evaluated_f * f(i);
        double root;
        if(mult < 0){
            root = 1;
        }else{
            root = 0;
        }
        row.push_back(i);
        row.push_back(f(i));
        row.push_back(mult);
        row.push_back(root);
        returnList.push_back(row);
        row.clear();
        evaluated_f = f(i);
    }
    return returnList;
}

int main()
{
    cout << "Enter start" << "\n";
    double start;
    cin >> start;

    cout << "Enter step" << "\n";
    double step;
    cin >> step;

    cout << "Enter end" << "\n";
    double end;
    cin >> end;

    vector<vector<double>> tabla = incrementalSearch(start, step, end);
    cout << "| " << 'x' 
            << " | " << "f(x)"
            << " | " << "f(a)*f(b)" 
            << " | " << "root \n";

    for (int i = 0; i < tabla.size(); i++)
    {
        cout << "| " << tabla[i][0] 
            << " | " << tabla[i][1] 
            << " | " << tabla[i][2] 
            << " | " << tabla[i][3] << "\n";
    }
}


